Content.Client/Singularity/Visualizers/RadiationCollectorSystem.cs

```diff
@@ -0,0 +1,105 @@
+using System;
+using Content.Shared.Singularity.Components;
+using Robust.Client.Animations;
+using Robust.Client.GameObjects;
+
+namespace Content.Client.Singularity.Visualizers;
+
+public sealed class RadiationCollectorSystem : VisualizerSystem<RadiationCollectorComponent>
+{
+    public override void Initialize()
+    {
+        base.Initialize();
+        SubscribeLocalEvent<RadiationCollectorComponent, ComponentInit>(OnComponentInit);
+        SubscribeLocalEvent<RadiationCollectorComponent, AnimationCompletedEvent>(OnAnimationCompleted);
+    }
+
+    private void OnComponentInit(EntityUid uid, RadiationCollectorComponent comp, ComponentInit args)
+    {
+        comp.ActivateAnimation = new Animation {
+            Length = TimeSpan.FromSeconds(0.8f),
+            AnimationTracks = {
+                new AnimationTrackSpriteFlick() {
+                    LayerKey = RadiationCollectorVisualLayers.Main,
+                    KeyFrames = {new AnimationTrackSpriteFlick.KeyFrame(comp.ActivatingState, 0f)}
+                }, // TODO: Make this play a sound when activating a radiation collector.
+            }
+        };
+
+        comp.DeactiveAnimation = new Animation {
+            Length = TimeSpan.FromSeconds(0.8f),
+            AnimationTracks = {
+                new AnimationTrackSpriteFlick() {
+                    LayerKey = RadiationCollectorVisualLayers.Main,
+                    KeyFrames = {new AnimationTrackSpriteFlick.KeyFrame(comp.DeactivatingState, 0f)}
+                }, // TODO: Make this play a sound when deactivating a radiation collector.
+            }
+        };
+    }
+
+    private void UpdateRadiationCollectorSpriteOrAnimation(EntityUid uid, RadiationCollectorVisualState state, RadiationCollectorComponent comp, SpriteComponent sprite, AnimationPlayerComponent? animPlayer = null)
+    {
+        if (state == comp.CurrentState)
+            return;
+
+        if ((RadiationCollectorVisualState) ((state ^ comp.CurrentState) & RadiationCollectorVisualState.Active) == RadiationCollectorVisualState.Active)
+            state = (RadiationCollectorVisualState) (state | RadiationCollectorVisualState.Deactivating); // Convert to transition state.
+
+        comp.CurrentState = state;
+
+        switch (state)
+        {
+            case RadiationCollectorVisualState.Activating:
+                if (!Resolve(uid, ref animPlayer, logMissing: false))
+                    goto case RadiationCollectorVisualState.Active;
+                AnimationSystem.Play(uid, animPlayer, comp.ActivateAnimation, RadiationCollectorComponent.AnimationKey);
+                break;
+            case RadiationCollectorVisualState.Deactivating:
+                if (!Resolve(uid, ref animPlayer, logMissing: false))
+                    goto case RadiationCollectorVisualState.Deactive;
+                AnimationSystem.Play(uid, animPlayer, comp.DeactiveAnimation, RadiationCollectorComponent.AnimationKey);
+                break;
```

Comment:
try to avoid using goto, also you should be resolving once and just make this vis only work with animplayer present